File: core/param_compat.py
from __future__ import annotations

from typing import Iterable


COMPATIBLE_FORM_FALLBACKS = {
    "b16": "fp16",
    "b32": "fp32",
}


def compatible_form(form: str | None) -> str | None:
    if form is None:
        return None
    return COMPATIBLE_FORM_FALLBACKS.get(str(form))


def form_candidates(form: str | None) -> list[str]:
    if form is None or str(form) == "":
        return []
    exact = str(form)
    compatible = compatible_form(exact)
    return [exact, compatible] if compatible and compatible != exact else [exact]


def pair_key_candidates(
    producer_op: str,
    producer_form: str,
    consumer_op: str,
    consumer_form: str,
) -> Iterable[tuple[str, str, bool]]:
    producer_forms = form_candidates(producer_form)
    consumer_forms = form_candidates(consumer_form)
    for p_form in producer_forms:
        for c_form in consumer_forms:
            used_compatible = p_form != producer_form or c_form != consumer_form
            yield f"{producer_op.upper()}.{p_form}", f"{consumer_op.upper()}.{c_form}", used_compatible

File: native/ParamCompat.cpp
// Copyright (c) 2026 Huawei Technologies Co., Ltd.
// SPDX-License-Identifier: CANN-1.0

#include "native/ParamCompat.h"

namespace vfsim::param_compat {
namespace {

std::string qualify(const std::string &op, const std::string &form) {
  return form.empty() ? op : op + "." + form;
}

} // namespace

std::string compatibleForm(const std::string &form) {
  if (form == "b16")
    return "fp16";
  if (form == "b32")
    return "fp32";
  return {};
}

std::vector<std::string> formCandidates(const std::string &form) {
  if (form.empty())
    return {};
  const std::string compatible = compatibleForm(form);
  if (!compatible.empty() && compatible != form)
    return {form, compatible};
  return {form};
}

std::vector<PairKeyCandidate>
pairKeyCandidates(const std::string &producerOp, const std::string &producerForm,
                  const std::string &consumerOp, const std::string &consumerForm) {
  std::vector<PairKeyCandidate> out;
  for (const std::string &pForm : formCandidates(producerForm)) {
    for (const std::string &cForm : formCandidates(consumerForm)) {
      out.push_back(PairKeyCandidate{
          qualify(producerOp, pForm),
          qualify(consumerOp, cForm),
          pForm != producerForm || cForm != consumerForm,
      });
    }
  }
  return out;
}

} // namespace vfsim::param_compat
